### ISSUE TITLE
Text Selection Issues with Japanese Text

### ISSUE THREAD

**Issue by cyphar** [koreader/koreader#8270]
* KOReader version: 2021.09
* Device: Android (Pixel 3a)

#### Issue

Because Japanese words are not auto-selected correctly by KOReader for dictionary lookup (see #4091) it is necessary to manually highlight the word I want to look up and then tap the dictionary button. This is not the end of the word in theory (and I understand not wanting to automatically use mecab on documents), but unfortunately it seems that there are a series of issues with making small selections which make using a dictionary with KOReader almost impossible. Here's a list of examples of text that cannot be selected properly:

 * A kanji followed by a single kana -- for instance in the text 本を読みながら, KOReader will not allow you to select *just* 読み (which is a word in of itself and due to  limitations in StarDict is how you would look up 読む because Japanese StarDict dictionaries have synonyms set up for verbs such that 読み is a synonym of 読む which is the dictionary form of the verb) but instead will force you to select 読みな which will not be correctly looked up in a dictionary.
 * Two kanji words (which is the vast **vast** majority of Japanese nouns) appear to also not be possible to select reliably. In 俺は男子でも, you cannot select 男子 by itself (instead you select 男子で which is not a word).
 * Trying to select two kana results in you selecting three, which is a problem for the same reason as above. 

The above two issues makes looking things up in the Japanese dictionary basically impossible. The core issue appears to be making small selections of two characters, but I've managed to do two character selections *sometimes* at random (it doesn't appear to be related to punctuation or where in a paragraph or sentence the selection is).

I assumed the issue was related to my font size, but even using the maximum font size doesn't solve the issue completely (though if you use a ridiculously large font size like 54 the problem almost goes away -- but that's not a practical size to use for reading actual books) and if you try to drag the selection back from three characters to two it basically always gets stuck on three. This issue is compounded by the fact that you cannot adjust a selection once you've made it.

##### Aside about Japanese Dictionaries and KOReader

Also just to clarify a question made in #4091:

> Now, with Chinese, Japanese and Korean (CJK), there is something everywhere in the code that decides that any such CJK char is itself a single word (and the HTML/CSS specs say the same). I don't know real much about these languages, and if really multiple CJK chars/ideographs can make out a single word. Can they? In general or only because there are entries with multiple chars in some dictionaries? Or in all dictionaries?

The vast majority of nouns in Japanese are two-or-more kanji words. Verbs and other forms of speech are also basically always more than one character as well (verbs are usually one kanji followed by one or more kana depending on how the verb has been inflected, but they could also be a series of kana if there is no kanji). The point is that there list of Japanese words which are only one character long is *incredibly* small in comparison to the rest of Japanese -- this means that anyone who has to do any lookup of any kind in Japanese text currently has to manually select the word and look it up by tapping the dictionary button. This is not an artefact of how dictionaries are organised, it's a fact of the language. It is my understanding that Chinese also has a lot of compound words and Korean has many multi-character words (though I have not studied either of those languages).

I also hasten to point out that in Japanese, fuzzy text searching is nowhere near as useful as with European languages because while European languages generally have rather long words where the conjugation only affects a few characters at the end of the word, in Japanese verbs are quite short (usually only one kanji and one kana) but the inflection of verbs can result in very long strings where the actual word is a small part of the conjugation -- and then the fuzzy searcher ends up removing the actual verb and just starts looking up words that look like the conjugation. For instance 喋られない is an inflection of 喋る -- the fuzzy searcher will end up looking for words that look like "られない" because it doesn't know that everything after 喋 is actually just Japanese grammatical inflection and thus you end up with completely useless lookups. I will open a separate issue to revive [this discussion](https://github.com/koreader/koreader/issues/4091#issuecomment-417635334) of having a hook for dictionary lookup so that we can have a Japanese deinflection plugin that makes lookups of Japanese verbs actually usable with KOReader. Japanese grammar is incredibly regular (there are only three exceptions in the entire language) so having such a plugin is remarkably straight-forward (Yomichan is a browser pop-up dictionary which does this incredibly well).

#### Steps to reproduce

1. Open any Japanese book.
2. Try to make a selection of only two adjacent characters.
3. Notice that your attempt will fail almost all of the time.

**Comment by poire-z** [comment]
I think that to switch from single word (hold) to multiple words selection (hold+pan), you need to move your finger by a distance large enough: the PAN_THRESHOLD set in there:
https://github.com/koreader/koreader/blob/a21b5be64a25f33d3d8194b8b1282805f9a990c2/frontend/device/gesturedetector.lua#L119-L127
You could try hardcoding a smaller value. (Read around in the file to see how these values are used.)
The normal user don't want something too small because otherwise, if he's just trembling a little bit (which may happen when you long press), single long-press would not activate and you would most always get hold+pan and the highlight dialog.
It's not currently configurable via the UI, but it could be if you feel other values are worth it and can help with some use cases.

As for the Japanese specific stuff, none of us active developpers read/write/understand these langauges, so we can't really solve an issue we can't really grasp.

From what I understand, I guess you could hook something around there, where long-press on the text is handled:
https://github.com/koreader/koreader/blob/a21b5be64a25f33d3d8194b8b1282805f9a990c2/frontend/apps/reader/modules/readerhighlight.lua#L836-L841
getWordFromPosition() is the engine (MuPDF, crengine) returning the "word" at your finger position. You don't want to hack that. But next, you could have a call to a Lua function that would check that "word": if it's a single char, depending on it's character class/script, do some additional stuff (or none for western languages): if a single Japanese glyph, look at next/prev char(s) to see the class/script and if they may form a compoud with the initial letter, and include them in the word. That may be possible with EPUBs where you have functions to grap prev/nextChar - not sure about PDFs. Might be a fun adventure, or hell :)
But again, none of use know about these languages, so we won't dare implementing anything...

**Comment by cyphar** [comment]
I think the question of making auto-selection of words should be handled separately (I sketched out an idea for how dictionary lookups of agglutinative languages could be improved in https://github.com/koreader/koreader/issues/4091#issuecomment-927510186) because there isn't really a trivial way to figure out whether you should select one or two characters given such a small amount of information -- that's something that's better left up to a proper text segmentation program.

But what if there was a way to adjust text selections after you made them (and when adjusting you get higher-precision selection). Something like this for instance (which would match the text selection UX on most mobile platforms):

![1024-2219](https://user-images.githubusercontent.com/2888411/134859458-9a0c6d40-a114-4695-910c-5ae3e6479253.png)

Though I think it would make more sense to put them on top since the default pop-up window is underneath the text by default.

Now this would in the short term still be somewhat annoying to use because it seems like it'd be necessary every time I need to look up a word, but if something like https://github.com/koreader/koreader/issues/4091#issuecomment-927510186 gets implemented (by me presumably) then you would only ever need the fine-grained text selection controls in cases where the text segmentation screwed up or you fat-fingered the selection.

**Comment by poire-z** [comment]
> But what if there was a way to adjust text selections after you made them (and when adjusting you get higher-precision selection). Something like this for instance (which would match the text selection UX on most mobile platforms):

Well, we don't have that kind of text selection, but we have another kind that you get AFTER you made some highlight, when you tap on this highlight, and it allows you to move the start or end of the current highlight (latest iteration at #7725).

May be we could have something like this happening when:
- long-press to select text
- for anything but CJK, the text selection spans up to prev/next space/punctuation: and it is usually more than one char => classic lookup
- for CJK, it normally results a single char
- when a single char is selected, propose this kind of dialog to allow extending the single char selection to grab chars around.
![image](https://user-images.githubusercontent.com/24273478/134975059-5d5c9b6c-fc70-4a4a-bf87-49dba6f71184.png)
(this is fake, I just made some highlights and removed/renamed some buttons :) just to show what I could get after I hit once the black button):

Westerners looking up `À` or `d'` or `ô` would get it too - but well, after 26 lookups, they should know these words :) and the result is just a tap away.

Would something like that help ?
No idea how simple/hard/messy it is to implement.

(And may be this dialog could have added buttons so CJK specific helper can plug themselves there to have more magic happening.)

**Comment by poire-z** [comment]
Btw, you haven't tell if/how [tweaking `PAN_THRESHOLD`  ](https://github.com/koreader/koreader/issues/8270#issuecomment-927349635) helps.

**Comment by cyphar** [comment]
> Btw, you haven't tell if/how [tweaking `PAN_THRESHOLD` ](https://github.com/koreader/koreader/issues/8270#issuecomment-927349635) helps.

Sorry, I've been trying to figure out how to test it on my (Android) device -- do I need to compile koreader and sideload it in order to make that tweak?

**Comment by poire-z** [comment]
> I would simply make this something that you can do when you create a selection not when you single-tap 

When you write "single-tap", you mean long-press without move, which launch dict lookup?
And "create selection", you mean long-press with move above this pan threshold to get the highlight dialog?
Wasn't the issue about small/single char text selection ?
I personally don't want to change anything to the current text selection way, which works well enough, we don't need to have an intermediate step (and to have it look like on a phone with these handles :)
Or maybe we can have the above too whenever the start or end char is CJK - where we can expect the user is having a hard time doing proper precise selection ?

> so that when I get #4091 working it won't interfere with that feature

Didn't you want us above to handle the 2 issues separately ? :) We're also helping Chinese and Korean readers in this one.
The interaction with #4091 would be when would it come to play, before or after this dialog.

> You could also make it conditional based on document language if you feel that it'd be too intrusive for folks reading non-CJK languages.

We can't trust document language tags, and we can have japanese bits embedded in an english or german book, so don't depend on stuff like this.
Btw, can you guess that a small fragment is Japanese and not Chinese by just the chars ? The kanji chars are shared between Japanese and Chinese, I think? Can you expect to always have some "little" kana specific to Japanese around ?

> Sorry, I've been trying to figure out how to test it on my (Android) device -- do I need to compile koreader and sideload it in order to make that tweak?

Compiling for Android is probably too much work. But we're Lua and our code stays text files.
So, if your phone is rooted, you can edit/update these files directly in `/data/data/org.koreader.launcher/files/` . Otherwise, there are ways to patch the code at startup, or have scripts executed to replace some files (which I'm not familiar with, https://github.com/koreader/koreader/wiki/Android-tips-and-tricks).

**Comment by cyphar** [comment]
> When you write "single-tap", you mean long-press without move, which launch dict lookup?

Yes, sorry I should've been more specific.

> Didn't you want us above to handle the 2 issues separately ? :) We're also helping Chinese and Korean readers in this one.

Yes, they should be handled separately. My point is more that if we had #4091 it wouldn't make sense to have this happen after the long-press-without-move because the dictionary lookup would work most of the time. But perhaps it'd be better to revisit how this point after #4091 gets solved :smile_cat:. I'll stop bringing that issue up. :sweat_smile: 

(Though since Korean has spaces I think they would probably prefer lookups to be done on the space boundary automatically just like with European languages -- but let's not go there right now. Better to ask some Korean learners what they'd prefer.)

> We can't trust document language tags, and we can have japanese bits embedded in an english or german book, so don't depend on stuff like this.

Ah, fair enough. Since the issue is that CJK characters are all interpreted as each word being a single character, maybe we should just make it so that the adjustable handles are available whenever you have a single-letter selection (maybe also requiring that the unicode character be in the CJK ranges or at least outside the common latin ranges?). That would be language-agnostic and shouldn't affect selections for languages where the existing selection system works well today.

> Btw, can you guess that a small fragment is Japanese and not Chinese by just the chars ? The kanji chars are shared between Japanese and Chinese, I think? Can you expect to always have some "little" kana specific to Japanese around ?

Very short answer to your question: Some of the time, but not always and it depends how many characters before/afterwards you want to look.

However since many Chinese words are also often two characters long, I don't think Chinese readers would mind have finer controls for selection. If Korean also currently is only selected a single character at a time rather than at a space boundary, I expect they'd also prefer it.

<details>
<summary>The slightly longer version.</summary>
Some words have identical kanji characters in both Chinese and Japanese, and even for words that are not shared there's no trivial way to detect that (you'd need to look it up in a dictionary and even then it might a false cognate or some other case like that). So unless the selection has Japanese-specific kana characters you can't really tell and that's not always guaranteed for every selection. For instance, in the sentence 彼は先生です if you selected just "先生" you wouldn't be able to tell whether it's Chinese or Japanese because the word exists in both languages -- you'd need to look for the はです characters. In addition, you can have grammatically-correct Japanese sentences that have no Japanese-specific kana at all in which case it's completely ambiguous. You could do the detection based on whether simplified characters are used but that won't help for non-Mainland-Mandarin books and probably has many other issues I'm not aware of since I don't study Chinese. On paper the same issue would exist with Korean Hanja but as far as I understand Hanja are not as common in modern text as they used to be.
</details>

> So, if your phone is rooted, you can edit/update these files directly in /data/data/org.koreader.launcher/files/ . Otherwise, there are ways to patch the code at startup, or have scripts executed to replace some files.

My phone isn't rooted sadly, but I'll try the script-based method (I saw that page but it wasn't clear to me whether it would be able to modify the Lua files the app uses). Thanks.

**Comment by cyphar** [comment]
I managed to get the sideloading working. Based on my testing a value of `self.PAN_THRESHOLD = self.DOUBLE_TAP_DISTANCE / 2` is pretty accurate and I can select individual characters now pretty accurately. However at really small font sizes `DOUBLE_TAP_DISTANCE / 3` is probably necessary. Maybe the `PAN_THRESHOLD` should be based on font size rather than being a hard-coded quantity?

For those interested, here is a script that lets you overwrite any part of the source tree on Android (it makes a backup of the source tree, but probably better to be careful when doing this).

```sh
#!/system/bin/sh

# $1 is the koreader user directory
# $2 is the koreader system directory

USER_DIR="$1"
SYSTEM_DIR="$2"

SIDELOAD_DIR="$USER_DIR/sideload"
if [ -d "$USER_DIR/sideload" ]; then
	# NOTE: This first copy can be very slow, so probably best to comment it out or only use this with scripts.afterupdate.
	cp -R "$SYSTEM_DIR" "$USER_DIR/sideload.old-$(date '+%s')-$$"
	cp -R "$SIDELOAD_DIR"/* "$SYSTEM_DIR"
fi
```

**Comment by cyphar** [comment]
>      

So, we're back on my proposal at #8270 (comment) ? Or do you think of something different?

**Comment by cyphar** [comment]
> So, we're back on my proposal at [#8270 (comment)](https://github.com/koreader/koreader/issues/8270#issuecomment-928227054) ? Or do you think of something different?

Ah, I misread that part of your proposal. Given that it would have to be restricted to avoid adding needless extra steps to the current flow (which I totally understand), I think having a tunable or self-adjustable PAN_THRESHOLD would be a better solution.

> > Based on my testing a value of self.PAN_THRESHOLD = self.DOUBLE_TAP_DISTANCE / 2 is pretty accurate and I can select individual characters now pretty accurately.
> 
> So, with this pan threshold tunable, would you still need the above "single char selection allows extending selection before lookup" ? Or would it then just be good enough ?

If the PAN_THRESHOLD was tunable (or self-adjusted based on text-size) the problem would no longer be big enough of an issue that you'd need to adjust things that way.

> > However at really small font sizes (16 or so) DOUBLE_TAP_DISTANCE / 3 is probably necessary. Maybe the PAN_THRESHOLD should be based on font size rather than being a hard-coded quantity?
> 
> The font size is only available in documents (and mostly in EPUB documents, with PDF, we only need to set a font size when reflow mode is used, so it might just not be set). We could override this default PAN_THRESHOLD with something related to the document font size when we open a book and change font size - and restore it back when going to File manager. Or just update it whenever we set/change a document font size and live with that until next time?

Maybe there should be some kind of sane default and if we're in a situation whether the font size is known we can adjust the PAN_THRESHOLD? Or maybe the PAN_THRESHOLD could be adjusted based on what kind of text the user is trying to select (as well as its font size if that's possible)? That way you would have more accurate PAN_THRESHOLD if you try to select CJK text, but you'd also have the less jittery PAN_THRESHOLD when trying to select non-CJK text.

I should probably try to open some English documents with a reduced PAN_THRESHOLD to see what kind of negative impact it has for non-CJK documents. I'll probably need to play around with how exactly the font size should affect the threshold.

**Comment by cyphar** [comment]
> But, given that a user is probably using overall the same "kind" of font size, or small variations around the one he is confortable with, may be just a static configurable pan_threshold in our Taps & gestures > Gestures intervals > Text selection pan distance (if a distance can be considered an interval in good english, as durations ?) would just be enough ? No need to make it more complex and variable across contexts.

Yeah that's probably the simplest solution. I'll send a PR. Thanks for your time. :smile_cat:

**Comment by poire-z** [comment]
^ Is this still required ? Or is Japanese text selection no more an issue with your recent Japanese support plugin ?
Can we close this issue?

**Comment by cyphar** [comment]
Yeah we can close this. There are some other issues with selection (even in the emulator, KOReader seems to offset the pointer position to the right enough to make it annoying to select the right starting character) but I'll open a separate issue/PR for that.
